### PR TITLE
PySide2 -> QtPy Migration (and Windows permissions error fixing)

### PR DESCRIPTION
**Pull request by finagle29**
This PR removes explicit dependencies on PySide2, replacing it with the QtPy abstraction layer library so that `pypeit_show_1dspec` works regardless of if the user has PySide2 or PyQt5 bindings installed.

Additionally, this PR fixes a Windows permissions error when updating Spec2DObj or AllSpec2DObj files (a file handle was still retained via the astropy `Table.read` call, copying the table means that the file is no longer in use after we have read it in so that deleting/overwriting it works smoothly in Windows)

Will post results of unit tests on Windows + macos once they finish.

### CODE REVIEW

**Comment by finagle29** [comment]
On Windows both test_runpypeit.py tests are failing, looks like the log file isn't being closed at the end of the `run_pypeit.main` function
`=========================== 2 failed, 251 passed, 547 warnings in 2202.40s (0:36:42) ============================`

**Comment by finagle29** [comment]
Reran test_runpypeit.py on Windows
```console
$ pytest pypeit\tests\test_runpypeit.py
...
================================== 2 passed, 237 warnings in 232.48s (0:03:52) ==================================
```

**Comment by finagle29** [comment]
and on macos `pypeit_show_2dspec` is the only test failing but I can't make out what the problem is from the stacktrace
`=========== 1 failed, 252 passed, 464 warnings in 1259.13s (0:20:59) ===========`

```
=================================== FAILURES ===================================
_______________________________ test_show_2dspec _______________________________

    @cooked_required
    def test_show_2dspec():
        droot = os.path.join(os.getenv('PYPEIT_DEV'), 'Cooked')
        spec2d_file = os.path.join(droot, 'Science',
                                 'spec2d_b27-J1217p3905_KASTb_2015May20T045733.560.fits')
        # Ginga needs to be open in RC mode
        display.connect_to_ginga(raise_err=True, allow_new=True)
        # Save
        cdir = os.getcwd()
        os.chdir(droot)
        # List
        pargs = show_2dspec.parse_args([spec2d_file, '--list'])
        show_2dspec.main(pargs)
        # Show
        pargs = show_2dspec.parse_args([spec2d_file])
>       show_2dspec.main(pargs)

/Users/milan/Documents/GitHub/PypeIt/pypeit/tests/test_scripts.py:163:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/Users/milan/Documents/GitHub/PypeIt/pypeit/scripts/show_2dspec.py:135: in main
    viewer, ch = display.show_image(image, chname=chname_skysub, waveimg=spec2DObj.waveimg, clear=True)
/Users/milan/Documents/GitHub/PypeIt/pypeit/display/display.py:175: in show_image
    out = canvas.clear()
/Users/milan/miniforge3/lib/python3.8/site-packages/ginga/util/grc.py:44: in _call
    return self._fn(self._chname, command, *args, **kwdargs)
/Users/milan/miniforge3/lib/python3.8/site-packages/ginga/util/grc.py:189: in call
    res = self._proxy.dispatch_call(method_name, p_args, p_kwdargs)
/Users/milan/miniforge3/lib/python3.8/xmlrpc/client.py:1109: in __call__
    return self.__send(self.__name, args)
/Users/milan/miniforge3/lib/python3.8/xmlrpc/client.py:1450: in __request
    response = self.__transport.request(
/Users/milan/miniforge3/lib/python3.8/xmlrpc/client.py:1153: in request
    return self.single_request(host, handler, request_body, verbose)
/Users/milan/miniforge3/lib/python3.8/xmlrpc/client.py:1169: in single_request
    return self.parse_response(resp)
/Users/milan/miniforge3/lib/python3.8/xmlrpc/client.py:1341: in parse_response
    return u.close()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <xmlrpc.client.Unmarshaller object at 0x7fdcc4a28cd0>

    def close(self):
        # return response tuple and target method
        if self._type is None or self._marks:
            raise ResponseError()
        if self._type == "fault":
>           raise Fault(**self._stack[0])
E           xmlrpc.client.Fault: <Fault 1: "<class 'KeyError'>:'sciimg-det01'">

/Users/milan/miniforge3/lib/python3.8/xmlrpc/client.py:655: Fault
----------------------------- Captured stdout call -----------------------------
Filename: /Users/milan/Documents/GitHub/PypeIt-development-suite/Cooked/Science/spec2d_b27-J1217p3905_KASTb_2015May20T045733.560.fits
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU     187   ()
  1  DET01-SCIIMG    1 ImageHDU        21   (350, 2048)   float32
  2  DET01-IVARRAW    1 ImageHDU        17   (350, 2048)   float32
  3  DET01-SKYMODEL    1 ImageHDU        17   (350, 2048)   float32
  4  DET01-OBJMODEL    1 ImageHDU        17   (350, 2048)   float32
  5  DET01-IVARMODEL    1 ImageHDU        17   (350, 2048)   float32
  6  DET01-TILTS    1 ImageHDU        17   (350, 2048)   float64
  7  DET01-SCALEIMG    1 ImageHDU        16   (1,)   float32
  8  DET01-WAVEIMG    1 ImageHDU        17   (350, 2048)   float64
  9  DET01-BPMMASK    1 ImageHDU        19   (350, 2048)   int16 (rescales to uint16)
 10  DET01-SLITS    1 BinTableHDU     48   1R x 10C   [1K, 2048D, 2048D, 2048D, 2048D, 2048D, 1I, 1I, 1D, 1D]
 11  DET01-SCI_SPEC_FLEXURE    1 BinTableHDU     22   1R x 2C   [K, D]
 12  DET01-DETECTOR    1 BinTableHDU     41   2R x 4C   [1D, 1D, 14A, 14A]
```

**Comment by finagle29** [comment]
merging develop in seems to have fixed that.
tests passing on macos
`================ 254 passed, 464 warnings in 1324.73s (0:22:04) ================`
and running on Windows

**Review by profxj — APPROVED** [review]
approved!
thanks

**Comment by finagle29** [comment]
tests passing on Windows!
`================================ 254 passed, 836 warnings in 2845.93s (0:47:25) ================================`

File: README.md
Comment (kbwestfall) [inline README.md:8]: Nice!  Can you also update/add to the shields in doc/index.rst?

File: doc/installing.rst
Comment (kbwestfall) [inline doc/installing.rst:27]: I'm not sure I follow how the installation works now in the sense that QtPy is a abstracting wrapper for the other packages.  Does pip install qtpy install one of these packages if none are installed on the system?  We might want to explain how this works in the installation docs.  The QtPy PyPI page says "You need PyQt5, PyQt4, PySide2 or PySide installed in your system to make use of QtPy.", and we might want to repeat something like this in the install doc.

**Review by kbwestfall — DISMISSED** [review]
Very nice.  Just some doc questions from me, so I'm approving now.

**Comment by jhennawi** [comment]
Hi @finagle29 this looks great. Before approving this I would like to request that you update the developer install documentation at doc/installing.rst. After the PySide2 dependency was added, these instructions were not updated and developers installs became impossible on some architectures. Please indicate exactly the steps now required for a developer install. 

After that I will approve and we can marge.

File: doc/installing.rst
Comment (tepickering) [inline doc/installing.rst:97]: last i checked `linetools` still required both PySide2 and PyQt5 to be installed in order to work.

Comment (jhennawi) [reply doc/installing.rst:97]: There was a recent PR to line tools: https://github.com/linetools/linetools/pull/526/files

Comment (finagle29) [reply doc/installing.rst:97]: Hmm could you let me know which version that was? PypeIt's requirements.txt was pinning an old linetools version (that required PyQt5) for quite a while (0.3dev2147) but a newer version of linetools (0.3dev2227) should require only PySide2 and the head of linetool's git repo should now only require QtPy. Once a new linetools version is released on pypi that only depends on QtPy, I'll update the language here.

But more generally, I probably need to take a deep look at the install and dev install docs and do a bit of rewriting.

Comment (jhennawi) [reply doc/installing.rst:97]: Thanks @finagle29. We definitely need to remove the PySide2 dependency implicit or explicit since that is at the heart of all the installation woes. If users want to use it and can handle installing it, then they can as per the ginga approach.

File: doc/installing.rst
Comment (finagle29) [reply doc/installing.rst:27]: That's a good point! I was thinking that there are generally 2 types of environments that users install PypeIt into:

1. a full environment with a whole bunch of stuff in it, some of which will probably require PyQt5 or PySide2, so the user will already have a required python Qt binding installed and ready to use
2. an empty environment in which case guis won't work. In this case we definitely need to provide guidance to users on what to do, but I think presenting users with a full list of requirements before quick and easy install instructions can be a little intimidating, so I think I'll change that when I do a dive into the install docs later (but still in this PR).

I also think its valuable to have a one-line install command presented at the top of the install docs for non-developer users who want PypeIt to Just Work. My goal with creating and maintaining the conda-forge recipe for PypeIt is to provide users with this experience, but I do know that there are varying opinions within the PypeIt team on conda as a dependency manager/all-in-one install tool, so I don't want to put forward `conda install -c conda-forge pypeit` as the recommended way to install PypeIt without greater buy-in, but that that might be a conversation for the Slack.

**Comment by profxj** [comment]
`linetools` pip updated:  https://pypi.org/project/linetools/

File: doc/installing.rst
Comment (tepickering) [reply doc/installing.rst:27]: as a long-time conda advocate, my opinion is that pure python packages like `pypeit` should use `pip` as the recommended installation method. it's fine to provide advice on how to use conda and provide some recipes, but i know from lived experience that pulling in the `conda-forge` channel *will* cause problems for many users. 

what i am working on setting up is the following:

- `pip install pypeit` will install the core dependencies, but not any of the gui tools or dependencies
- `pip install pypeit[gui]` will install the gui dependencies (`ginga`, `linetools`, and `qtpy`) along with the scripts that use them
- `pip install pypeit[gui,pyside2]` or `pip install [gui,pyqt5]` will be provided as a convenience for pulling in the user's choice of QT binding if they don't already have it.

this would give the user more control over what they pull in based on their needs. iirc, conda doesn't have the same provision for such optional dependencies.

**Comment by jhennawi** [comment]
@finagle29 did you get the show2d_spec thing working. That is a serious problem if it persists. It worked on my Mac, but I didn't not run that test, although I can.

**Comment by finagle29** [comment]
@jhennawi yes, merging in develop fixed the show_2dspec bug

File: doc/installing.rst
Comment (jhennawi) [inline doc/installing.rst:144]: @finagle29 this is different from the procedure I had to follow to get this to work, specifically the yaml file containing the dependencies which get installed is not mentioned here. Specifically I had to run: 

conda env create -f environment.yml 
conda activate pypeit 
pip install pyqt5
conda activate pypeit 
pip install -e .

Can you please add a note on this. You can state that if the users prefer to install the dependencies themselves via conda or pip they can also do that, but the yaml file should be mentioned here. 

Thanks!

Comment (finagle29) [reply doc/installing.rst:151]: I think I addressed this in my latest commit, could you take a look at it?

**Review by profxj — APPROVED** [review]
all good by me now.

**Review by jhennawi — APPROVED** [review]
Looks great. Merging.

**Pull request merged**
